src/Microsoft.DotNet.Watcher.Tools/Internal/ProcessRunner.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.Watcher.Internal
+{
+    public class ProcessRunner
+    {
+        private readonly ILogger _logger;
+
+        public ProcessRunner(ILoggerFactory loggerFactory)
+        {
+            Ensure.NotNull(loggerFactory, nameof(loggerFactory));
+
+            _logger = loggerFactory.CreateLogger<ProcessRunner>();
+        }
+
+        public async Task<int> RunAsync(ProcessSpec processSpec, CancellationToken cancellationToken)
+        {
+            Ensure.NotNull(processSpec, nameof(processSpec));
+            Ensure.NotNull(cancellationToken, nameof(cancellationToken));
+
+            var process = CreateProcess(processSpec);
+            var tcs = new TaskCompletionSource<object>();
+            EventHandler handler = (sender, args) =>
+            {
+                tcs.TrySetResult(null);
+            };
+
+            process.Exited += handler;
+            
+            cancellationToken.Register(() =>
+            {
+                process?.KillTree();
```

Comment:
actually process.HasExited would not work. Maybe assign to a local variable first and then do `.?` ?